### ISSUE TITLE
Support CSL YAML (and or JSON) bibliographies

### ISSUE THREAD

**Issue by daniel-eder** [typst/typst#2924]
### Description

Currently typst can use BIB and Hayagriva-style YAML bibliography files. Considering that CSL style YAML/JSON bibliographies have been around for a while and have widespread support in all de-facto standard tools (e.g. Zotero for CSL JSON, BetterBibTex for CSL YAML and JSON) support would seem desirable.

Further, typst already uses CSL styles, so it makes sense to support the standard end-to-end, including bibliography files (without conversion via bib)

BetterBibTex has a feature request to support Hayagriva-style YAML, however the author argued (correctly, in my opinion) that CSL is a now many years-old standard that warrants support by typst: https://github.com/retorquere/zotero-better-bibtex/issues/2711 

If you disagree it would be interesting to collect arguments to find out on which side which additional format should be supported to not be limited to BIB across tools.

### Use Case

Users of bibliography management software that supports CSL can use typst without converting to bib as intermediate format, and avoid issues associated with that. 

See https://citationstyles.org/ for a list of software that uses CSL (by the way, this list should probably feature typst, considering the style side of CSL is already implemented :) )

**Comment by DerDrodt** [comment]
The arguments in the linked issue seem sensible. As someone who was not involved with early development at all: What was the initial reason to create a new format in hayagriva? Why not use CSL?

As I see it, there are two options here:
1. Add CSL bibliography support to hayagriva _directly_, i.e., without any translation, at which point the hayagriva format will likely slowly fade away.
2. Add support for translating hayagriva <-> CSL. Going through Bib(La)TeX is not ideal, and translation would greatly expand typst's use cases.

Since I don't know how attached the core dev team is to hayagriva as a format, I would appreciate their input here. I am open to implementing (2), and to support development of (1).

**Comment by reknih** [comment]
Hayagriva actually supports CSL-JSON, but only for applying the `citeproc.js` test suite. Structs are generic in such a way that native Hayagriva entries and CSL-JSON cannot be mixed.

While I think that it would be worthwhile to enable this mixing of entry formats, I also think that the native Hayagriva format has its place. CSL-JSON was designed to expose all CSL variables and switches in Citeproc.js (look at e.g. dates) instead of ease of entering and reading by humans. Both Hayagriva and BibTeX are harder to parse, but, at least in the case of the former, this is by design. I aimed for a compact and readable representation when creating the format, which, I believe, JSON generally and CSL-JSON in particular do not offer.

**Comment by HiramTheHero** [comment]
Any updates on this?

I'm running into multiple issues trying to have typst format my bibliography correctly. (Much akin to the use case @daniel-eder pointed out.) I believe this is from .bib files not being exported with the correct info for typst to translate to CSL in order to format the bibliography properly. I assume having CSL YAML or JSON support would ensure that the bibliographies look identical to what Zotero's CSL bibliography exports look like. This would be a much better experience than wrestling with bib file exports or trying to reformat Zotero CSL YAML exports to be Hayagriva compatible. 

In the meantime, if anyone knows of a way to convert CSL YAML/JSON to Hayagriva YAML, I would appreciate any pointers. (From my searching, it seems like there is virtually no software support for the Hayagriva YAML file format.)

**Comment by ezwelty** [comment]
I'm impressed by Typst – in particular, that writing in multi scripts just ... works – but puzzled over the need for yet a new citation style.

> CSL-JSON was designed to expose all CSL variables and switches in citeproc.js (look at e.g. dates) instead of ease of entering and reading by humans. Both Hayagriva and BibTeX are harder to parse, but, at least in the case of the former, this is by design. I aimed for a compact and readable representation when creating the format, which, I believe, JSON generally and CSL-JSON in particular do not offer.

@reknih I agree with you that the ergonomics of JSON (at least in a plain text editor without syntax highlighting and realtime validation) are not as friendly as YAML for editing and reading by humans. But what about CSL YAML? Also, how common is it nowadays for bibliography files to be maintained by hand instead of being exported from e.g. Zotero?

**Comment by ironupiwada** [comment]
@reknih I agree with the voices against Hayagriva. It's still a non-functional format that can't be converted to any other one (unlike the BibTeX/biblatex referenced here) and with which UTF8 collation can't be sorted correctly (which is crucial for at least multilingual bibliographic entries). Especially since many of the problems reported against Hayagriva's operation remain unanswered (such as the absence of support for prefixes in footnotes); [I've been waiting for nine months](https://github.com/typst/hayagriva/issues/193) to get a simple 'yes/no' answer whether non-English entries will be properly sorted). To me, in such a situation, the only thing left is to convert the document to Markdown and work with Quarto and Zotero.

**Comment by laurmaedje** [comment]
As for bugs and incompleteness in the bibliography subsystem: That's fairly unrelated to the hayagriva format. As @Drodt mentions, CSL is very complex and there is simply no fully compliant CSL processor written in Rust.

As for CSL-JSON: We are open to supporting it as a bibliography input format.

**Comment by erauchway** [comment]
For folks in the humanities or others who need to cite archival sources, a JSON bibliography file is essential. It would be great to be able to use Typst for anything that required citations but right now it seems like there's no way to do that.

**Comment by fronesis47** [comment]
Joining this conversation because I’ve recently discovered a bunch of really simply, but annoying bugs in hayagriva, along with bugs in the Typst `#bibliography()` function, even when using `.bib` files (I’m guessing these are related). I have realized I can’t really use it to process academic bibliographies.

Maybe others have already seen this, but I developed the following workaround:
1. Export CSL JSON (from Zotero for me).
2. Use pandoc with `—citeproc` to generate to output to the biblio entries  as `.typ`. The trick here is not to use the standalone `-s` flag, so that…
3. In your typst document, you can load that `.typ` file directly at the end, with `#include`.

Maybe I’ve missed something, but in my testing, pandoc is working flawlessly with the CSL JSON to create `.typ` entries.  Obviously the pandoc conversion is an extra step, but it’s worth it to still get the features of Typst without the bibliography bugs.

**Comment by retorquere** [comment]
The bibtex - CSL route involves [unnecessary risk of information loss](https://retorque.re/zotero-better-bibtex/exporting/pandoc/index.html), so I would argue it would serve the user best to at least offer a straight CSL path. It seems to me that until citeproc-csl is replaced in typst, any format other than raw CSL-JSON carries this unnecessary risk.

**Comment by Drodt** [comment]
typst/hayagriva#489 allows direct support for CSL-JSON. It's work in progress, but planned.

**Comment by retorquere** [comment]
> [hayagriva] can't be converted to any other one

BBT offers preliminary support for Hayagriva import/export since the newest release.

**Comment by v4u6h4n** [comment]
Thanks for adding Hayagriva! Decided to migrate to Typst yesterday so couldn't be better timed :-)

**Comment by retorquere** [comment]
It's all very fresh, and I'm not a typst user myself, so I welcome feedback on the implementation.

**Comment by fronesis47** [comment]
I tested the BBT Hayagriva export with a small zotero collection of 4 items, that I had built to demonstrate the bugs in hayagriva/typst. These are problematic types for hayagriva, including "thesis," "conference paper," and "book section." 

But it looks like they remain problematic for BBT Hayagriva. This is probably still a problem with Hayagriva, but perhaps it's helpful to list the issues I saw with the BBT Hayagriva exports, when then rendered in Typst with the `#bibliography()` function:

- For thesis – drops the University and "PhD Dissertation."
- For conference paper – drops the name of the conference. 
- For book section with multiple authors - displays the chapter title as the book title and drops the actual book title. 

I'm comparing this to what happens if I just generate a bibliography directly with zotero. 

I really appreciate @retorquere 's heroic efforts, both with BBT in general and with this Hayagriva support. But unless I'm missing something, Hayagriva is still so broken that one is still better off using some other workaround (I listed my own workaround earlier in this thread).  Here's a link to the [hayagriva bug I filed](https://github.com/typst/hayagriva/issues/466), which is still a bug with the latest releases of typst and hayagriva.

**Comment by retorquere** [comment]
The current implementation is naive -- I don't use typst, so what I have now is cribbed from the one page I could find on the hayagriva site (which I've since lost) and an existing gist implementation I found. If something needs to change, open an issue on the BBT project, include a `refs` debug log (which gives me a source item for the test case) and the expected output (part 2 of the test case) and I'll make the change.

**Comment by fronesis47** [comment]
> The current implementation is naive -- I don't use typst, so what I have now is cribbed from the one page I could find on the hayagriva site (which I've since lost) and an existing gist implementation I found. If something needs to change, open an issue on the BBT project, include a `refs` debug log (which gives me a source item for the test case) and the expected output (part 2 of the test case) and I'll make the change.

Sure. I’ll open an issue and try provide some helpful info.

**Comment by v4u6h4n** [comment]
I have ~4000 items in my Zotero, and ran into some initial errors, but they all seem to have been incorrectly formatted dates and language codes (I must have been mashing my keyboard when entering them!), which I just corrected in Zotero, and since then everything seems to be working fine so far.

**Comment by retorquere** [comment]
My exports have a very permissive date parser/reformatter, so if you have samples of those dates, please report them in a BBT github issue.

**Comment by fronesis47** [comment]
Sorry it took me so long, but I've now opened an issue at the BBT GitHub project, with the proper zotero debug ID and the list of issues.
